### ISSUE TITLE
Make automatic adding missing imports smarter

### ISSUE THREAD

**Issue by chriskuech** [Microsoft/TypeScript#29878]
### Scenario 1: Fuzzy Logic
I have a several files containing `import {connect} from "react-redux";`.  If I use the add missing imports context menu to add the missing import for `connect`, the system should realize "react-redux" is far more probable module than "net", "tls", and "http2", which are all recommended higher than "react-redux", but none of which are used anywhere in my project.

### Scenario 2: Respect node conventions
If I have a file `./myfolder/index.ts`, I can import the file with `import "./myfolder"`.  The system currently resolves missing imports from `./myfolder` as `./myfolder/index.ts`, when it should actually use `./myfolder`.

**Comment by mjbvz** [comment]
@minestarks Scenario 1 sounds like it could fall under IntelliCode

Scenario two sounds like #29835, likely a configuration issue. @chriskuech Please share your tsconfig

**Comment by chriskuech** [comment]
This does sound like a duplicate issue

```jsonc
{
  "compilerOptions": {
    "target": "es5",
    "downlevelIteration": true,
    "allowJs": true,
    "skipLibCheck": false,
    "esModuleInterop": true,
    "allowSyntheticDefaultImports": true,
    "strict": true,
    "forceConsistentCasingInFileNames": true,
    "module": "esnext",
    "moduleResolution": "node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "jsx": "preserve",
    "lib": ["dom", "es7"],
    "rootDirs": [ // not working on last check
      "./src"
    ]
  },
  "include": [
    "src"
  ]
}
```

**Comment by minestarks** [comment]
@chriskuech 

> the system should realize "react-redux" is far more probable module than "net", "tls", and "http2"

What context is available for us to determine the probability? Are you suggesting the list be sorted on something like the popularity of libraries, or based on something particular about your project?

**Comment by chriskuech** [comment]
Yes I would ideally like the list to be sorted first by number of occurrences in my project.  I have accidentally added the wrong `connect` too many times to count.  Do you happen to know how the list is currently sorted?  The current order ("net", "tls", and "http2", "react-redux") does not seem to be alphabetical or any other obvious order.

**Comment by minestarks** [comment]
Sorry, I don't know off the top of my head - might well be arbitrary at the moment.  @mjbvz Scenario 1 seems more in the domain of TS itself rather than IntelliCode as it could be implemented pretty deterministically without machine learning.
